### PR TITLE
[SPARK-28477] [SQL] Rewrite CaseWhen with single branch to If

### PR DESCRIPTION
**Pull request by davidvrba**
### What changes were proposed in this pull request?
Spark org.apache.spark.sql.functions do not have `if` function so conditions are expressed using `when-otherwise` function. However `If` (which is available in SQL) has more efficient code gen. This pr rewrites `when-otherwise` conditions to `If` if it is possible (`when-otherwise` with single branch)

### Why are the changes needed?
It is an optimization enhancement. Here is a simple performance comparison (tested in local mode (with 4 cores)):
```
val df = spark.range(10000000000L).withColumn("x", rand)
val resultA = df.withColumn("r", when($"x" < 0.5, lit(1)).otherwise(lit(0))).agg(sum($"r"))
val resultB = df.withColumn("r", expr("if(x < 0.5, 1, 0)")).agg(sum($"r"))

resultA.collect() // takes 56s to finish
resultB.collect() // takes 30s to finish
```
### Does this PR introduce any user-facing change?
No

### How was this patch tested?
New test is added.

### CODE REVIEW

**Review by srowen — COMMENTED** [review]
Seems plausible, @cloud-fan ?

**Review by HyukjinKwon — COMMENTED** [review]
@davidvrba can you show some perf number since its optimization?

**Comment by davidvrba** [comment]
@HyukjinKwon sure, i edited the description with one simple performance comparison (it is tested in local mode).

**Comment by maropu** [comment]
I tried in the current master though, I couldn't see much difference... (do I miss something?);
```
$ java -version
java version "1.8.0_181"
Java(TM) SE Runtime Environment (build 1.8.0_181-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.181-b13, mixed mode)
$ ./bin/spark-shell --master=local[1] --conf spark.driver.memory=8g --conf spark.sql.shuffle.partitions=1 -v

scala> val df = spark.range(10000000000L)
scala> val whenVer = df.withColumn("r", when($"id" % lit(2) === lit(0), lit(1)).otherwise(lit(0))).agg(sum($"r"))
scala> val ifVer = df.withColumn("r", expr("if(id % 2 = 0, 1, 0)")).agg(sum($"r"))

scala> timer { whenVer.write.format("noop").mode("overwrite").save() }
Elapsed time: 15.130525497s                                                     

scala> timer { ifVer.write.format("noop").mode("overwrite").save() }
Elapsed time: 13.953640182s  
```
```
// Benchmarks w/o aggregates
scala> val df = spark.range(400000000L)
scala> val whenVer = df.withColumn("r", when($"id" % lit(2) === lit(0), lit(1)).otherwise(lit(0)))
scala> val ifVer = df.withColumn("r", expr("if(id % 2 = 0, 1, 0)"))

scala> timer { whenVer.write.format("noop").mode("overwrite").save() }
Elapsed time: 8.498007927s                                                      

scala> timer { ifVer.write.format("noop").mode("overwrite").save() }
Elapsed time: 8.375774959s                                                      
```
FYI: the generated code for them;
```
// whenVer
...
/* 035 */     byte project_caseWhenResultState_0 = -1;
/* 036 */     do {
/* 037 */       boolean project_isNull_3 = true;
/* 038 */       boolean project_value_3 = false;
/* 039 */       boolean project_isNull_4 = false;
/* 040 */       long project_value_4 = -1L;
/* 041 */       if (2L == 0) {
/* 042 */         project_isNull_4 = true;
/* 043 */       } else {
/* 044 */         project_value_4 = (long)(project_expr_0_0 % 2L);
/* 045 */       }
/* 046 */       if (!project_isNull_4) {
/* 047 */         project_isNull_3 = false; // resultCode could change nullability.
/* 048 */         project_value_3 = project_value_4 == 0L;
/* 049 */
/* 050 */       }
/* 051 */       if (!project_isNull_3 && project_value_3) {
/* 052 */         project_caseWhenResultState_0 = (byte)(false ? 1 : 0);
/* 053 */         project_project_value_2_0 = 1;
/* 054 */         continue;
/* 055 */       }
/* 056 */
/* 057 */       project_caseWhenResultState_0 = (byte)(false ? 1 : 0);
/* 058 */       project_project_value_2_0 = 0;
/* 059 */
/* 060 */     } while (false);
/* 061 */     // TRUE if any condition is met and the result is null, or no any condition is met.
/* 062 */     final boolean project_isNull_2 = (project_caseWhenResultState_0 != 0);
...

// ifVer
...
/* 034 */     boolean project_isNull_3 = true;
/* 035 */     boolean project_value_3 = false;
/* 036 */     boolean project_isNull_4 = false;
/* 037 */     long project_value_4 = -1L;
/* 038 */     if (2L == 0) {
/* 039 */       project_isNull_4 = true;
/* 040 */     } else {
/* 041 */       project_value_4 = (long)(project_expr_0_0 % 2L);
/* 042 */     }
/* 043 */     if (!project_isNull_4) {
/* 044 */       project_isNull_3 = false; // resultCode could change nullability.
/* 045 */       project_value_3 = project_value_4 == 0L;
/* 046 */
/* 047 */     }
/* 048 */     boolean project_isNull_2 = false;
/* 049 */     int project_value_2 = -1;
/* 050 */     if (!project_isNull_3 && project_value_3) {
/* 051 */       project_isNull_2 = false;
/* 052 */       project_value_2 = 1;
/* 053 */     } else {
/* 054 */       project_isNull_2 = false;
/* 055 */       project_value_2 = 0;
/* 056 */     }
...
```

**Comment by davidvrba** [comment]
@maropu  I rerun the query (from the description) again (in different order) and got slightly more conservative result: 52s for query with `when` and 40s for query with `if`. 

I also run your query also in current master (now on different laptop with 8 cores): 

```
val df = spark.range(10000000000L)
val ifVer = df.withColumn("r", expr("if(id % 2 = 0, 1, 0)")).agg(sum($"r"))
val whenVer = df.withColumn("r", when($"id" % lit(2) === lit(0), lit(1)).otherwise(lit(0))).agg(sum($"r"))

spark.time(ifVer.write.format("noop").mode("overwrite").save())
spark.time(whenVer.write.format("noop").mode("overwrite").save())
```
And I run the `save` 10 times and compute average `Time taken` and get 3980.3ms for `ifVer` vs 4630.6ms for `whenVer` which leads to 14% reduction in time.

@JoshRosen Could you please also provide some performance benchmarks for this?

**Comment by srowen** [comment]
Looks like it saves about a microsecond. Pretty trivial but a win's a win? given there is very little extra complexity. @maropu is it possibly a win here that the generated code is a little smaller? maybe it helps when this is part of a large plan?

**Comment by HyukjinKwon** [comment]
One thing we might have to consider is, analysing time. When the plan is deeply nested or has too many columns, this time overwhelms time to time. We might have to avoid to add a rule for a trivial improvement I guess. I don't feel strongly if a committer likes it but otherwise I wouldn't add.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 1]
```diff
@@ -435,6 +435,11 @@ object SimplifyConditionals extends Rule[LogicalPlan] with PredicateHelper {
           e.copy(branches = newBranches)
         }
 
+      case e @ CaseWhen(branches, elseValue) if branches.length == 1 =>
```
[the remark this answers is not in the document]
Comment (davidvrba) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: Thanks for looking at it! And yes, it does depend on how codegen is implemented. I will move the logic to `CaseWhen.doGenCode` as you suggest.

**Comment by maropu** [comment]
> Looks like it saves about a microsecond. Pretty trivial but a win's a win? given there is very little extra complexity. @maropu is it possibly a win here that the generated code is a little smaller? maybe it helps when this is part of a large plan?

Yea, that might be so. As shown above, this optimization can makes the generated code a bit more simpler, so IMO JVM JIT might be able to optimize it easily and it has no harm on performance. But, the gain seems to be trivial, so I don't have a strong preference.

**Comment by cloud-fan** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala [hunk 2]
```diff
@@ -186,98 +186,106 @@ case class CaseWhen(
...
-      ctx.addMutableState(CodeGenerator.javaType(dataType), ev.value),
-      dataType)
+    if (branches.length == 1) {
+      // If we have only single branch we can use If expression and its codeGen
+      If(
+        branches(0)._1,
+        branches(0)._2,
+        elseValue.getOrElse(Literal.create(null, branches(0)._2.dataType))).doGenCode(ctx, ev)
+    } else {
+      // This variable holds the state of the result:
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala]: we can create a method `multiBranchesCodegen` to put the original code:
```
override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
  if ...
  else {
    multiBranchesCodegen
  }
}
```
then the diff would be much smaller.

Comment (davidvrba) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala]: yes, it makes sense

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala [hunk 3]
```diff
@@ -185,7 +185,7 @@ case class CaseWhen(
     "CASE" + cases + elseCase + " END"
   }
 
-  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+  def multiBranchesCodegen(ctx: CodegenContext, ev: ExprCode): ExprCode = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala]: nit: private

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
retest this please

**Comment by davidvrba** [comment]
@cloud-fan Do you have any suggestion how to deal with the failing unit test? Can you please correct me: My understanding is that the test requires number of global variables to be equal to 1, but `If` expression generates 0, this is why the test fails with `ArrayBuffer() had size 0 instead of expected size 1`. So we could just modify the condition in the unit test as `ctx.inlinedMutableStates.size =< 1`. Is that correct, or do i get it wrong?

**Comment by cloud-fan** [comment]
can we slightly update that test to use 2 branches?

**Review by maropu — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by davidvrba** [comment]
Thank you very much!

**Pull request closed without merging**